File: py/solved/58.py

def prime(n):
	if n < 2:
            return False
	if n == 2:
            return True
	elif n%2 == 0:
            return False
	else:
            i = 3
            while i < n**.5 + 1:
                if n%i == 0:
                    return False
                i += 2
            return True


def solve():
    ratio = .1
    d = 0
    n = 1
    p = 0.0

    while 1:
        d += 2
        for i in range(3):
            n += d
            if prime(n): p += 1.0
        n += d
        print(p/(2*d + 1))
        if p/(2.*(d+1) - 1) < ratio:
            break	
    print(d + 1, p)
    print('\n', str(int(100*p/(2*(d+1) - 1))) + "%")

from timer import time_function
print(time_function(solve))

File: cpp/solved/58.cpp
#include <iostream>
#include <chrono>
#include <cmath>
#include "helper/prime.h"

using namespace std;

int solve() {
    int n, s = 3, i, N = 7;
    double p = 0.0;
    for (n = 2, i = 1; ; n++, i++) {
        if (i%(s-1) == 0 && prime(n)) {
            p += 1.0;
        }
        //cout << "n = " << n << " i = " << i << " p = " << p << " s = " << s << " r = " << p << "/" << (2*s-1) << " = " << p/(2*s-1) << endl;
        if (n%2 == 1 && sqrt(n) == ((int)sqrt(n))) {
            if (p/(2*s+1) < .1) {
                break;
            }
            i = 0;
            s += 2;
        }
    }
    return s;
}

int main() {
    typedef chrono::high_resolution_clock Clock;
    typedef chrono::milliseconds milliseconds;
    Clock::time_point start = Clock::now();
    cout << "Answer: " << solve() << endl;
    Clock::time_point end = Clock::now();
    milliseconds ms = chrono::duration_cast<milliseconds>(end-start);
    cout << "Time: " << ms.count() << " ms" << endl;
    return 0;
}
